File: depertin_cliente/functions/tipos_entrega.js
// Arquivo: functions/tipos_entrega.js
//
// Tabela canônica de tipos de entrega aceitos por uma loja. Espelha (com
// mesma API semântica) o arquivo Dart em:
//   - depertin_cliente/lib/constants/tipos_entrega.dart
//   - depertin_web/lib/constants/tipos_entrega.dart
//
// Mantenha as três em sincronia. Builds separadas impedem import comum.

"use strict";

const COD = Object.freeze({
    BICICLETA: "bicicleta",
    MOTO: "moto",
    CARRO: "carro",
    CARRO_FRETE: "carro_frete",
});

const ORDEM_CANONICA = [COD.BICICLETA, COD.MOTO, COD.CARRO, COD.CARRO_FRETE];

const HIERARQUIA = Object.freeze({
    [COD.BICICLETA]: 1,
    [COD.MOTO]: 2,
    [COD.CARRO]: 3,
    [COD.CARRO_FRETE]: 4,
});

const TABELA_FRETE_POR_TIPO = Object.freeze({
    [COD.BICICLETA]: "padrao",
    [COD.MOTO]: "padrao",
    [COD.CARRO]: "carro",
    [COD.CARRO_FRETE]: "carro_frete",
});

const CADEIA_FALLBACK_TABELA = Object.freeze({
    [COD.CARRO_FRETE]: ["carro_frete", "carro", "padrao"],
    [COD.CARRO]: ["carro", "padrao"],
    [COD.MOTO]: ["padrao"],
    [COD.BICICLETA]: ["padrao"],
});

const RAIO_KM_RECOMENDADO = Object.freeze({
    [COD.BICICLETA]: 2.0,
    [COD.MOTO]: 15.0,
    [COD.CARRO]: 25.0,
    [COD.CARRO_FRETE]: 100.0,
});

/**
 * Normaliza o tipo de veículo do entregador (campo `veiculoTipo`, `veiculo`
 * ou `tipo_veiculo` em `users/{uid}`) para o set canônico.
 * Aceita variações: "Moto", "moto", "motocicleta", "Carro", "carro popular",
 * "carro_frete", "carro frete", "fiorino", "pick-up", "kombi", "bicicleta",
 * "bike".
 */
function normalizarTipoVeiculo(raw) {
    const s = String(raw || "").trim().toLowerCase();
    if (!s) return "";
    if (s.includes("frete") || s.includes("fiorino") || s.includes("pick") ||
        s.includes("kombi") || s.includes("utilitário") || s.includes("utilitario") ||
        s.includes("van") || s === "carro_frete") {
        return COD.CARRO_FRETE;
    }
    if (s.includes("carro")) return COD.CARRO;
    if (s.includes("moto") || s.includes("scooter") || s.includes("motocicleta")) {
        return COD.MOTO;
    }
    if (s.includes("bike") || s.includes("bicicleta") || s.includes("bicy")) {
        return COD.BICICLETA;
    }
    return "";
}

/**
 * Normaliza a lista `tipos_entrega_permitidos` de uma loja. Descarta valores
 * fora do set canônico, remove duplicatas, ordena por hierarquia crescente.
 */
function normalizarLista(raw) {
    if (!Array.isArray(raw)) return [];
    const set = new Set();
    for (const v of raw) {
        const s = String(v || "").trim().toLowerCase();
        if (HIERARQUIA[s] != null) set.add(s);
    }
    return Array.from(set).sort(
        (a, b) => (HIERARQUIA[a] || 0) - (HIERARQUIA[b] || 0),
    );
}

/** Tipo de MAIOR hierarquia da lista (ou null se vazia). */
function maiorTipo(tipos) {
    if (!Array.isArray(tipos) || tipos.length === 0) return null;
    let melhor = null;
    let maior = -1;
    for (const t of tipos) {
        const h = HIERARQUIA[t] || -1;
        if (h > maior) {
            maior = h;
            melhor = t;
        }
    }
    return melhor;
}

/**
 * Retorna true se o entregador pode receber a corrida.
 * `aceitos` vazio → loja legada sem config → NÃO filtra (compat).
 */
function compativel(tipoVeiculoEntregador, aceitos) {
    if (!Array.isArray(aceitos) || aceitos.length === 0) return true;
    return aceitos.includes(tipoVeiculoEntregador);
}

/**
 * Deriva default para lojas legado. Chamar só quando o doc da loja ainda
 * não tem `tipos_entrega_permitidos`. Input:
 *   - `temProdutoRequerVeiculoGrande`: bool (agregado dos produtos da loja).
 */
function defaultLegado({ temProdutoRequerVeiculoGrande }) {
    if (temProdutoRequerVeiculoGrande) {
        return [COD.CARRO, COD.CARRO_FRETE];
    }
    // Política conservadora — preserva comportamento pré-migração para
    // lojas legado. Bicicleta e carro são opt-in explícito do lojista.
    return [COD.MOTO];
}

function lerDeDoc(data) {
    if (!data || typeof data !== "object") return [];
    return normalizarLista(data.tipos_entrega_permitidos);
}

/**
 * Normaliza a categoria **escolhida pelo lojista no momento da solicitação**
 * (`pedido.tipo_entrega_solicitado`). Diferente de `tipos_entrega_permitidos`
 * (lista aceita pela loja), aqui o valor é uma string canônica única
 * decidida a cada clique em "Solicitar entregador". Aceita variações e
 * devolve código canônico (ou string vazia se inválido).
 */
function normalizarTipoSolicitado(raw) {
    const s = String(raw || "").trim().toLowerCase();
    if (!s) return "";
    if (HIERARQUIA[s] != null) return s;
    return normalizarTipoVeiculo(s);
}

/**
 * Deriva a categoria efetivamente buscada para o pedido.
 *
 * Ordem de precedência:
 *   1. `pedido.tipo_entrega_solicitado` — escolha explícita do lojista.
 *   2. Se a lista aceita tem **apenas um** tipo → esse tipo é implícito.
 *   3. `null` quando a loja aceita múltiplos tipos mas ainda não escolheu
 *      (bloqueia o despacho até nova decisão).
 *   4. `null` também para loja legado sem config (caller decide se filtra
 *      ou não — compatibilidade preservada em outro helper).
 *
 * Retornos possíveis:
 *   - string canônica (ex: "moto") → filtra entregadores por esse tipo exato.
 *   - null → não há categoria decidida para este pedido.
 */
function categoriaEfetivaPedido(pedido, tiposAceitosLoja) {
    const explicito = normalizarTipoSolicitado(pedido && pedido.tipo_entrega_solicitado);
    if (explicito) return explicito;
    const aceitos = Array.isArray(tiposAceitosLoja)
        ? normalizarLista(tiposAceitosLoja)
        : [];
    if (aceitos.length === 1) return aceitos[0];
    return null;
}

function rotulo(codigo) {
    switch (codigo) {
        case COD.BICICLETA: return "Bicicleta";
        case COD.MOTO: return "Moto";
        case COD.CARRO: return "Carro popular";
        case COD.CARRO_FRETE: return "Carro frete";
        default: return String(codigo || "");
    }
}

module.exports = {
    COD,
    TIPOS: COD,
    ORDEM_CANONICA,
    HIERARQUIA,
    TABELA_FRETE_POR_TIPO,
    CADEIA_FALLBACK_TABELA,
    RAIO_KM_RECOMENDADO,
    normalizarTipoVeiculo,
    normalizarLista,
    maiorTipo,
    compativel,
    defaultLegado,
    lerDeDoc,
    rotulo,
    normalizarTipoSolicitado,
    categoriaEfetivaPedido,
};

File: depertin_cliente/lib/constants/tipos_entrega.dart
// Arquivo: lib/constants/tipos_entrega.dart
//
// Tipos canônicos de veículo/entrega aceitos por uma loja. Cada loja decide
// quais tipos são compatíveis com seus produtos; o carrinho calcula o frete
// usando o tipo de MAIOR hierarquia aceito pela loja (maior = mais caro), e
// o backend filtra entregadores pelo tipo do veículo ativo deles.
//
// Regras de hierarquia:
//   bicicleta (1) < moto (2) < carro (3) < carro_frete (4)
//
// Exemplos:
//   - Loja aceita ["bicicleta","moto","carro"]  → frete calculado como CARRO
//   - Loja aceita ["moto","carro"]              → frete calculado como CARRO
//   - Loja aceita ["carro_frete"]               → frete calculado como CARRO_FRETE
//   - Loja aceita ["bicicleta"]                 → frete calculado como BICICLETA
//                                                 (usa tabela `padrao` com
//                                                 aviso de limite de ~2 km)
//
// Cliente NUNCA escolhe o tipo de veículo — isso é decisão interna do
// sistema a partir da config da loja.

import 'package:cloud_firestore/cloud_firestore.dart';

/// Enum canônico dos 4 tipos aceitos pelo sistema.
enum TipoEntrega {
  bicicleta,
  moto,
  carro,
  carroFrete,
}

/// Helpers estáticos e tabela de hierarquia.
class TiposEntrega {
  TiposEntrega._();

  /// Código persistido no Firestore (snake_case estável).
  static const String codBicicleta = 'bicicleta';
  static const String codMoto = 'moto';
  static const String codCarro = 'carro';
  static const String codCarroFrete = 'carro_frete';

  /// Ordem canônica para listagem em UI (bike → moto → carro → carro frete).
  static const List<String> ordemCanonica = <String>[
    codBicicleta,
    codMoto,
    codCarro,
    codCarroFrete,
  ];

  /// Hierarquia logística. Maior = frete mais caro / carga maior / veículo
  /// mais robusto. Usado pra decidir a tabela de frete quando a loja aceita
  /// mais de um tipo.
  static const Map<String, int> hierarquia = <String, int>{
    codBicicleta: 1,
    codMoto: 2,
    codCarro: 3,
    codCarroFrete: 4,
  };

  /// Tabela `tabela_fretes` a consultar. `bicicleta` e `moto` compartilham
  /// `padrao`. Se futuramente quiser tabela própria de bike, basta trocar.
  static const Map<String, String> tabelaFretePorTipo = <String, String>{
    codBicicleta: 'padrao',
    codMoto: 'padrao',
    codCarro: 'carro',
    codCarroFrete: 'carro_frete',
  };

  /// Cadeia de fallback se a tabela primária não existir para a cidade.
  /// Ordem: preferida primeiro. Se nenhuma for encontrada, o cálculo cai
  /// em `_taxaBaseFallback` da UI.
  static const Map<String, List<String>> cadeiaFallbackTabela =
      <String, List<String>>{
    codCarroFrete: <String>['carro_frete', 'carro', 'padrao'],
    codCarro: <String>['carro', 'padrao'],
    codMoto: <String>['padrao'],
    codBicicleta: <String>['padrao'],
  };

  /// Raio km recomendado (aviso na UI; NUNCA bloqueia compra por distância).
  static const Map<String, double> raioKmRecomendado = <String, double>{
    codBicicleta: 2.0,
    codMoto: 15.0,
    codCarro: 25.0,
    codCarroFrete: 100.0,
  };

  /// Rótulo amigável para UI.
  static String rotulo(String codigo) {
    switch (codigo) {
      case codBicicleta:
        return 'Bicicleta';
      case codMoto:
        return 'Moto';
      case codCarro:
        return 'Carro popular';
      case codCarroFrete:
        return 'Carro frete';
      default:
        return codigo;
    }
  }

  /// Descrição curta pra card/checkbox no painel do lojista.
  static String descricaoCurta(String codigo) {
    switch (codigo) {
      case codBicicleta:
        return 'Entregas pequenas até ~2 km da loja.';
      case codMoto:
        return 'Ideal para refeições, encomendas leves.';
      case codCarro:
        return 'Carros de passeio (volumes médios).';
      case codCarroFrete:
        return 'Utilitários: Fiorino, Montana, pick-ups, Kombi.';
      default:
        return '';
    }
  }

  /// Normaliza lista vinda do Firestore:
  /// - descarta valores fora do set canônico
  /// - remove duplicatas
  /// - ordena por hierarquia ascendente (bike→moto→carro→frete)
  static List<String> normalizarLista(dynamic raw) {
    if (raw is! Iterable) return const <String>[];
    final set = <String>{};
    for (final v in raw) {
      final s = v?.toString().trim().toLowerCase();
      if (s != null && hierarquia.containsKey(s)) set.add(s);
    }
    final lista = set.toList()
      ..sort((a, b) => (hierarquia[a] ?? 0).compareTo(hierarquia[b] ?? 0));
    return lista;
  }

  /// Retorna o tipo de MAIOR hierarquia da lista. Usado pra decidir a
  /// tabela de frete. Se lista vazia, retorna `null` (caller decide
  /// fallback para default legado).
  static String? maiorTipoDaLista(List<String> tipos) {
    if (tipos.isEmpty) return null;
    String? melhor;
    var maior = -1;
    for (final t in tipos) {
      final h = hierarquia[t] ?? -1;
      if (h > maior) {
        maior = h;
        melhor = t;
      }
    }
    return melhor;
  }

  /// Verifica se o veículo do entregador é compatível com a lista de tipos
  /// aceitos pela loja. Usado no backend pra filtrar a fila de despacho.
  static bool compativel(String tipoVeiculoEntregador, List<String> aceitos) {
    if (aceitos.isEmpty) return true; // loja legada sem config → não filtra
    return aceitos.contains(tipoVeiculoEntregador);
  }

  /// Normaliza um rótulo livre de veículo (ex: `"Moto"`, `"Carro de Frete"`,
  /// `"Fiorino"`, `"bike"`) para um código canônico do set. Replica a
  /// heurística do backend (`functions/tipos_entrega.js#normalizarTipoVeiculo`)
  /// para manter 100% de paridade entre o filtro client-side do radar e o
  /// filtro server-side do despacho. Retorna string vazia se não reconhecer.
  static String normalizarTipoVeiculo(dynamic raw) {
    final s = (raw?.toString() ?? '').trim().toLowerCase();
    if (s.isEmpty) return '';
    if (s.contains('frete') ||
        s.contains('fiorino') ||
        s.contains('pick') ||
        s.contains('kombi') ||
        s.contains('utilitário') ||
        s.contains('utilitario') ||
        s.contains('van') ||
        s == codCarroFrete) {
      return codCarroFrete;
    }
    if (s.contains('carro')) return codCarro;
    if (s.contains('moto') ||
        s.contains('scooter') ||
        s.contains('motocicleta')) {
      return codMoto;
    }
    if (s.contains('bike') || s.contains('bicicleta') || s.contains('bicy')) {
      return codBicicleta;
    }
    return '';
  }

  /// Deriva default para lojas legado que ainda não configuraram o campo.
  ///
  /// Política conservadora (abr/2026): preserva o comportamento de cobrança
  /// anterior à migração, pra lojas legado não surpreenderem clientes com
  /// aumento súbito de frete.
  ///
  /// - `temProdutoRequerVeiculoGrande=true` → `["carro","carro_frete"]`
  ///   (produto volumoso existia no carrinho; antes já calculava como
  ///   "carro", agora mantém e abre porta para `carro_frete` via fallback
  ///   na tabela).
  /// - `false` → `["moto"]` apenas. Bicicleta e carro são **opt-in
  ///   explícito** do lojista — não entram em default.
  static List<String> defaultLegado({
    required bool temProdutoRequerVeiculoGrande,
  }) {
    if (temProdutoRequerVeiculoGrande) {
      return const <String>[codCarro, codCarroFrete];
    }
    return const <String>[codMoto];
  }

  /// Converte para `List<String>` pronto para gravar no Firestore.
  static List<String> paraFirestore(List<String> tipos) =>
      normalizarLista(tipos);

  /// Lê da estrutura de documento (mapa ou snapshot.data()). Retorna lista
  /// vazia se o campo não existir.
  static List<String> lerDeDoc(
    Map<String, dynamic>? data, {
    String campo = 'tipos_entrega_permitidos',
  }) {
    if (data == null) return const <String>[];
    return normalizarLista(data[campo]);
  }

  /// Normaliza `tipo_entrega_solicitado` — a categoria **escolhida pelo
  /// lojista no momento do clique** em "Solicitar entregador". Diferente
  /// de [lerDeDoc], aqui o campo é uma string única (não lista). Aceita
  /// código canônico direto ou variações livres (ex: "Carro Popular",
  /// "Fiorino"); devolve string vazia se não reconhecer.
  static String normalizarTipoSolicitado(dynamic raw) {
    final s = (raw?.toString() ?? '').trim().toLowerCase();
    if (s.isEmpty) return '';
    if (hierarquia.containsKey(s)) return s;
    return normalizarTipoVeiculo(s);
  }

  /// Categoria efetivamente buscada para o pedido.
  ///
  /// Ordem de precedência:
  ///   1. `pedido.tipo_entrega_solicitado` — escolha explícita do lojista.
  ///   2. Lista aceita com **apenas um tipo** → tipo implícito.
  ///   3. `null` quando a loja aceita múltiplos tipos e ainda não houve
  ///      decisão (bloqueia despacho) ou quando a loja é legado sem config.
  static String? categoriaEfetivaPedido(
    Map<String, dynamic>? pedido,
    List<String>? tiposAceitosLoja,
  ) {
    final explicito = normalizarTipoSolicitado(
      pedido == null ? null : pedido['tipo_entrega_solicitado'],
    );
    if (explicito.isNotEmpty) return explicito;
    final aceitos = tiposAceitosLoja == null
        ? const <String>[]
        : normalizarLista(tiposAceitosLoja);
    if (aceitos.length == 1) return aceitos.first;
    return null;
  }
}

/// Container com snapshot carregado do Firestore — facilita passar entre
/// camadas (cart_screen, checkout, pedido).
class TiposEntregaLoja {
  TiposEntregaLoja({
    required this.lojaId,
    required this.tiposAceitos,
    required this.atualizadoEm,
  });

  final String lojaId;
  final List<String> tiposAceitos;
  final Timestamp? atualizadoEm;

  /// Atalho pra saber o alvo de frete.
  String? get maiorTipo => TiposEntrega.maiorTipoDaLista(tiposAceitos);

  factory TiposEntregaLoja.deDoc(
    String lojaId,
    Map<String, dynamic>? data,
  ) {
    if (data == null) {
      return TiposEntregaLoja(
        lojaId: lojaId,
        tiposAceitos: const <String>[],
        atualizadoEm: null,
      );
    }
    return TiposEntregaLoja(
      lojaId: lojaId,
      tiposAceitos: TiposEntrega.lerDeDoc(data),
      atualizadoEm: data['tipos_entrega_atualizado_em'] as Timestamp?,
    );
  }

  bool get estaConfigurada => tiposAceitos.isNotEmpty;
}
